### PR TITLE
Enable CUDA graph for ADAM optimizer

### PR DESCRIPTION
**Pull request by vasunvidia**
Add wait stream before copying next batch to CG input

Add OptimizerCudaGraphWrapper to CUDA graph optimizer

Cleanup

# What does this PR do ?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by asolergi-nv** [comment]
/ok to test [5284819](https://github.com/NVIDIA/Megatron-LM/pull/3429/commits/5284819766324e4408f159d9702e4ec7a381a35c)

File: megatron/core/optimizer/optimizer_cuda_graph.py
Comment (jaredcasper) [inline megatron/core/optimizer/optimizer_cuda_graph.py:1]: Shouldn't this file live in `megatron/core/optimizer/`?

File: megatron/core/optimizer_cuda_graph.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+# Copyright (c) 2024, NVIDIA CORPORATION. All rights reserved.
+
+"""Full iteration CUDA graph for training."""
```
Comment (jaredcasper) [inline megatron/core/optimizer_cuda_graph.py]: This description doesn't make much sense given the name of the file.

File: megatron/core/optimizer/optimizer_config.py [hunk 2]
```diff
@@ -355,6 +355,14 @@ class OptimizerConfig:
     config_logger_dir: str = ""
     """When non-empty, dumps entry-point configs to config_logger_dir"""
 
+    on_device_clip_grad: bool = False
```
Comment (jaredcasper) [inline megatron/core/optimizer/optimizer_config.py]: Is there a downside to doing this? Should we just default to on, or even just do it with no knob?

File: megatron/training/training.py [hunk 3]
```diff
@@ -3223,6 +3226,9 @@ def trace_handler(p):
     if args.cuda_graph_impl == "transformer_engine" and cuda_graph_helper.graphs_created():
         cuda_graph_helper.delete_cuda_graphs()
 
+    if args.optimizer_cuda_graph:
+        del optimizer.step
```
Comment (maanug-nv) [inline megatron/training/training.py:3214]: seems a bit unintuitive , can we just add a comment explaining that it resets it back from the wrapper to the normal method impl ?

**Review by maanug-nv — APPROVED** [review]
LGTM. left 1 nit. approving to unblock, but please add the comment before merging.

**Comment by ko3n1g** [comment]
/ok to test [131c546](https://github.com/NVIDIA/Megatron-LM/pull/3429/commits/131c546b21506c293f92339fb1bf2757e242a0c2)

File: megatron/core/optimizer/optimizer_cuda_graph.py
Comment (vasunvidia) [reply megatron/core/optimizer/optimizer_cuda_graph.py:1]: Moved the file.

File: megatron/core/optimizer/optimizer_config.py (hunk 2 above)
Comment (vasunvidia) [reply megatron/core/optimizer/optimizer_config.py]: Added a commit to remove on_device_clip_grad knob and use on_device_clip_grad if the kernel is present.

File: megatron/core/optimizer_cuda_graph.py (hunk 1 above)
Comment (vasunvidia) [reply megatron/core/optimizer_cuda_graph.py]: Fixed it.

File: megatron/core/optimizer/optimizer_cuda_graph.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
+# Copyright (c) 2024, NVIDIA CORPORATION. All rights reserved.
```
Comment (jaredcasper) [inline megatron/core/optimizer/optimizer_cuda_graph.py]: Wrong year.

File: megatron/core/optimizer/__init__.py [hunk 5]
```diff
@@ -528,6 +528,7 @@ def _get_megatron_optimizer_based_on_param_groups(
                 "weight_decay": config.weight_decay,
                 "betas": (config.adam_beta1, config.adam_beta2),
                 "eps": config.adam_eps,
+                "capturable": config.optimizer_cuda_graph,
```
Comment (jaredcasper) [inline megatron/core/optimizer/__init__.py:551]: Do we know when this argument was added to Adam? Does this need a version check?

Comment (vasunvidia) [reply megatron/core/optimizer/__init__.py:551]: This was added 2 years back. Version check is not needed.

File: megatron/training/training.py (hunk 3 above)
Comment (vasunvidia) [reply megatron/training/training.py:3214]: Added comment.

File: megatron/core/optimizer/optimizer_cuda_graph.py (hunk 4 above)
Comment (vasunvidia) [reply megatron/core/optimizer/optimizer_cuda_graph.py]: Fixed.

**Review by jaredcasper — APPROVED** [review]

**Review by ericharper — APPROVED** [review]
